### PR TITLE
remove k8s.io/kubernetes dependency from yurtctl join/reset

### LINKED ISSUE #671: [feature request]Refactor yurtctl join to remove k8s.io/kubernetes dependency
**What would you like to be added**:
`yurtctl join` references `k8s.io/kubernetes` dependencies, especially `k8s.io/kubernetes/cmd/kubeadm`, maybe we should refactor to remove the dependencies.

`yurtctl join` can be divided into the following steps: 
1) preflightcheck: check the necessary configuration for cluster operation, such as the pull of necessary images, the installation of kubelet.
2) start kubelet
3) start yurthub

**Why is this needed**:
We can remove the dependency of `k8s.io/kubernetes` in `yurtctl join` .(issue: #550 )

**others**
/kind feature
/kind good-first-issue

### PR DESCRIPTION
**Pull request by rambohe-ch**
/kind enhancement

#### What this PR does / why we need it:
Remove k8s.io/kubernetes dependency from `yurtctl join` and `yurtctl reset`.  

##### 1. feature update points:

   - In order to reduce the complexity of `yurtctl join` command and integrate with [sealer](https://github.com/alibaba/sealer), join `ControlPlane` node feature is removed, so `yurtctl join` can only join cloud or edge worker node. and end user should use `yurtctl init` to install `ControlPlane` through `sealer`.
   - `yurtctl reset` can only reset worker node that joined by `yurtctl join` command. if end user use `yurtctl reset` to reset other kind of nodes, the result maybe unknown.

##### 2.  yurtctl join command

-  implementation update points:
    - dependency on kubeadm copied from `k8s.io/kubernetes/cmd/kubeadm` to `github.com/openyurtio/openyurt/pkg/yurtctl/kubernetes`. so `KubeletConfiguration` has been left.

   - command line parameter:
        - In order to reduce the complexity of input parameter, `kubeadmapi.JoinConfiguration` and `kubeadmapi.InitConfiguration` have been removed. only command line parameters are supported in `yurtctl join`. 
       - add `YurtJoinData` interface.
       - `apiServerAddress` and `token` parameter for `yurtctl join` command
       - `organizations` parameter is added for adding customized `Organizations` info into yurthub client certificate.
       - `node-labels` parameter is added for adding customized `labels` for worker node.

  - prepare phase(pkg/yurtctl/cmd/join/phases/prepage.go):
       - kubelet service conf file(/etc/systemd/system/kubelet.service.d/10-kubeadm.conf) is the same of cloud and edge worker node, so merge `EdgeKubeletUnitConfig ` and `CloudKubeletUnitConfig ` into `KubeletUnitConfig`
       - add `/etc/kubernetes/kubelet.conf` file preparation.
       - add `/etc/kubernetes/pki/ca.crt` file preparation.

  -  preflight phase(pkg/yurtctl/cmd/join/phases/preflight.go)
       - add `/etc/kubernetes/kubelet.conf` and `/etc/kubernetes/pki/ca.crt` file existing check
       - add `pause` and `yurthub` image pull check

  - joinNode phase(pkg/yurtctl/cmd/join/phases/joinnode.go)
       - merge `joinCloudNode/joindEdgeNode` into `joinNode` because the join routine is the same except the `WorkingMode` parameter of yurthub.
       - `openyurt.io/is-edge-worker` label is added as kubelet command line parameter, so after kubelet startup, only `CRISocket` annotation is need to patch to node.
      - because kubelet use `HTTP` to connect yurthub, so `rotate-certificates=faluse` is added to kubelet command line parameter.
      - use `templates.SubsituteTemplate` to populate yurthub yaml, so the format of fields in `YurthubTemplate` are modified. for example: from `__workingMode__` to `{{.workingMode}}`

  - postCheck phase(pkg/yurtctl/cmd/join/phases/postcheck.go)
      - remove nodeLabel patch.

#####  3. yurtctl reset command
  -  add `resetData` interface
  - add preflight phase
  - separate `cleanfile` phase into `cleanupnode` and `cleanyurtfile` phases.

#### Which issue(s) this PR fixes:

Fixes #671 

#### Does this PR introduce a user-facing change?

```release-note

```

### CODE REVIEW

**Comment by rambohe-ch** [comment]
@adamzhoul @Peeknut @zyjhtangtang

**Comment by rambohe-ch** [comment]
@neo502721

**Comment by adamzhoul** [comment]
hi  @rambohe-ch 
I see this is a big and important update. we may take a while to fully understand.
can you have a simple description of what you did? so can help us understand more.
something like :
1. copy k8s code to dir `yurtctl/kubernetes`
2. merge joinCloudNode/joindEdgeNode into one( and I see you simplify the implantation can you have some introduction ？)
3. ...

PS:  can some change in node can be implantation in node-servant? to make the update to kubelet/yurthub  keep same.

very thanks.

**Comment by adamzhoul** [comment]
hi  @rambohe-ch 
I see this is a big and important update. we may take a while to fully understand.
can you have a simple description of what you did? so can help us understand more.
something like :
1. copy k8s code to dir `yurtctl/kubernetes`
2. merge joinCloudNode/joindEdgeNode into one( and I see you simplify the implantation can you have some introduction ？)
3. ...

PS:  can some change in node can be implantation in node-servant? to make the update to kubelet/yurthub  keep same.

very thanks.

**Comment by rambohe-ch** [comment]
> hi @rambohe-ch I see this is a big and important update. we may take a while to fully understand. can you have a simple description of what you did? so can help us understand more. something like :
> 
> 1. copy k8s code to dir `yurtctl/kubernetes`
> 2. merge joinCloudNode/joindEdgeNode into one( and I see you simplify the implantation can you have some introduction ？)
> 3. ...
> 
> PS: can some change in node can be implantation in node-servant? to make the update to kubelet/yurthub keep same.
> 
> very thanks.

@adamzhoul Thanks for your suggestions. i 

> hi @rambohe-ch I see this is a big and important update. we may take a while to fully understand. can you have a simple description of what you did? so can help us understand more. something like :
> 
> 1. copy k8s code to dir `yurtctl/kubernetes`
> 2. merge joinCloudNode/joindEdgeNode into one( and I see you simplify the implantation can you have some introduction ？)
> 3. ...
> 
> PS: can some change in node can be implantation in node-servant? to make the update to kubelet/yurthub keep same.
> 
> very thanks.

@adamzhoul Thanks for your suggestions. i have updated the pull request messages.

File: pkg/yurtctl/cmd/reset/phases/preflight.go [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+// runPreflight executes preflight checks logic.
+func runPreflight(c workflow.RunData) error {
+	r, ok := c.(resetData)
+	if !ok {
+		return errors.New("preflight phase invoked with an invalid data struct")
+	}
+
+	if !r.ForceReset() {
+		fmt.Println("[reset] WARNING: Changes made to this host by 'kubeadm init' or 'kubeadm join' will be reverted.")
```
Comment (adamzhoul) [inline pkg/yurtctl/cmd/reset/phases/preflight.go:53]: why fmt not klog?

Comment (rambohe-ch) [reply pkg/yurtctl/cmd/reset/phases/preflight.go]: fixed

**Pull request merged**